vm/os-windows.cpp

```diff
@@ -304,6 +304,58 @@ void handle_ctrl_c() {
   SetConsoleCtrlHandler(factor::ctrl_handler, TRUE);
 }
 
+static volatile bool stop_on_ctrl_break = false;
+static HANDLE ctrl_break_thread = NULL;
+const int ctrl_break_sleep = 30; /* msec */
+
+static DWORD WINAPI ctrl_break_thread_proc(LPVOID mainThread) {
+  bool ctrl_break_handled = false;
+  while (factor::stop_on_ctrl_break) {
+    if (GetAsyncKeyState(VK_CANCEL) >= 0) { /* Ctrl-Break is released. */
+      ctrl_break_handled = false;  /* Wait for the next press. */
+    } else if (!ctrl_break_handled) {
+      /* This proc runs in its own thread without stopping the main
+         thread. Since in practice nobody uses the multi-VM stuff yet,
+         we just interrupt the first VM we can get. */
+      FACTOR_ASSERT(thread_vms.size() > 0);
+      THREADHANDLE thd = thread_vms.begin()->first;
+      factor_vm* vm = thread_vms.begin()->second;
+      /* Check if the VM thread has the same Id as the thread Id of the
+         currently active window. Note that thread Id is not a handle. */
+      DWORD fg_thd_id = GetWindowThreadProcessId(GetForegroundWindow(), NULL);
+      if ((fg_thd_id == vm->thread_id) && !vm->fep_p) {
+        atomic::store(&vm->skip_debugger_p, true);
+        vm->safepoint.enqueue_fep(vm);
+        wake_up_thread(thd); /* Try to wake up the thread. */
+        ctrl_break_handled = true;
+      }
+    }
+    Sleep(ctrl_break_sleep);
+  }
+  return 0;
+}
+
+void factor_vm::primitive_disable_ctrl_break() {
+  stop_on_ctrl_break = false;
+  if (ctrl_break_thread != NULL) {
+    DWORD wait_result = WaitForSingleObject(ctrl_break_thread,
+                                            2 * ctrl_break_sleep);
+    if (wait_result != WAIT_OBJECT_0)
+      TerminateThread(ctrl_break_thread, 0);
+    CloseHandle(ctrl_break_thread);
+    ctrl_break_thread = NULL;
+  }
+}
+
+void factor_vm::primitive_enable_ctrl_break() {
+  stop_on_ctrl_break = true;
+  if (ctrl_break_thread == NULL) {
+    DisableProcessWindowsGhosting();
+    ctrl_break_thread = CreateThread(NULL, 0, factor::ctrl_break_thread_proc,
+                                     NULL, 0, NULL);
+  }
```

Comment:
If you change the 4th parameter to CreateThread to `static_cast<LPVOID>(this)`, then `ctrl_break_thread_proc` would get a reference to the running vm as its first argument.